library/Booster/Syntax/Json/Internalise.hs

```diff
@@ -269,41 +311,111 @@ internaliseTerm ::
     Except PatternError Internal.Term
 internaliseTerm = internaliseTermRaw IsNotQQ
 
--- Throws errors when a term is encountered. The 'And' case
--- is analysed before, this function produces an 'AndPredicate'.
-internalisePredicate ::
+{- | Internalises an And-ed set of predicates, classifying them into
+  BoolPred, CeilPred, SubstitutionPred, and UnsupportedPred.
+  The substitution (as a whole) is analysed after internalisation, to
+  ensure nothing is circular or ambiguous.
+
+  Throws a PatternError when any of the top-level patterns is a term
+  or an And with a term (must be analysed beforehand).
+-}
+internalisePredicates ::
```

Comment:
:exploding_head:

Reply:
Added a description to `mkSubstitution` and a link to the gist